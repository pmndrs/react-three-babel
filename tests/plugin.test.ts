import * as babel from "@babel/core";
import plugin from "../src/index";

const example = `
import { createRoot } from '@react-three/fiber'

createRoot(canvasNode).render(
  <mesh>
    <boxGeometry />
    <meshStandardMaterial />
  </mesh>
)
`;

it("works", () => {
  const { code } = babel.transform(example, {
    plugins: [plugin],
    sourceType: "module",
  })!;
  expect(code).toMatchSnapshot();
});

const exampleWithDrei = `
import { createRoot } from '@react-three/fiber'
import { Html } from 'drei'

createRoot(canvasNode).render(
  <mesh>
    <boxGeometry />
    <meshStandardMaterial />
    <Html>
      <div>Hello World</div>
    </Html>
  </mesh>
)
`;

it("works with drei", () => {
  const { code } = babel.transform(exampleWithDrei, {
    plugins: [plugin],
    sourceType: "module",
  })!;
  expect(code).toMatchSnapshot();
});

const exampleWithImportShadowing = `
import { createRoot } from '@react-three/fiber'
import { BoxGeometry } from 'three'

createRoot(canvasNode).render(
  <mesh>
    <boxGeometry />
    <meshStandardMaterial />
  </mesh>
)
`;

it("works with import shadowing", () => {
  const { code } = babel.transform(exampleWithImportShadowing, {
    plugins: [plugin],
    sourceType: "module",
  })!;
  expect(code).toMatchSnapshot();
});

const exampleWithThreeStdLib = `
import { createRoot } from '@react-three/fiber'

createRoot(canvasNode).render(
  <mesh>
    <boxGeometry />
    <meshStandardMaterial />
  </mesh>
)
`;

it("allows customization of the import source", () => {
  const { code } = babel.transform(exampleWithThreeStdLib, {
    plugins: [
      [
        plugin,
        {
          importSource: "three-stdlib",
        },
      ],
    ],
    sourceType: "module",
  })!;
  expect(code).toMatchSnapshot();
});

const exampleWithExtend = `
import { createRoot, extend } from '@react-three/fiber'

createRoot(canvasNode).render(
  <mesh>
    <boxGeometry />
    <meshStandardMaterial />
  </mesh>
)
`;

it("does not break if extend is already imported", () => {
  const { code } = babel.transform(exampleWithExtend, {
    plugins: [plugin],
    sourceType: "module",
  })!;
  expect(code).toMatchSnapshot();
});
