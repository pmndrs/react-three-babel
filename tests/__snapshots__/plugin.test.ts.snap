// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`allows customization of the import source 1`] = `
"import { MeshStandardMaterial as _MeshStandardMaterial } from \\"three-stdlib\\";
import { BoxGeometry as _BoxGeometry } from \\"three-stdlib\\";
import { Mesh as _Mesh } from \\"three-stdlib\\";
import { extend as _extend } from \\"@react-three/fiber\\";
import { createRoot } from '@react-three/fiber';

_extend({
  Mesh: _Mesh,
  BoxGeometry: _BoxGeometry,
  MeshStandardMaterial: _MeshStandardMaterial
})

createRoot(canvasNode).render(<mesh>
    <boxGeometry />
    <meshStandardMaterial />
  </mesh>);"
`;

exports[`does not break if extend is already imported 1`] = `
"import { MeshStandardMaterial as _MeshStandardMaterial } from \\"three\\";
import { BoxGeometry as _BoxGeometry } from \\"three\\";
import { Mesh as _Mesh } from \\"three\\";
import { extend as _extend } from \\"@react-three/fiber\\";
import { createRoot, extend } from '@react-three/fiber';

_extend({
  Mesh: _Mesh,
  BoxGeometry: _BoxGeometry,
  MeshStandardMaterial: _MeshStandardMaterial
})

createRoot(canvasNode).render(<mesh>
    <boxGeometry />
    <meshStandardMaterial />
  </mesh>);"
`;

exports[`works 1`] = `
"import { MeshStandardMaterial as _MeshStandardMaterial } from \\"three\\";
import { BoxGeometry as _BoxGeometry } from \\"three\\";
import { Mesh as _Mesh } from \\"three\\";
import { extend as _extend } from \\"@react-three/fiber\\";
import { createRoot } from '@react-three/fiber';

_extend({
  Mesh: _Mesh,
  BoxGeometry: _BoxGeometry,
  MeshStandardMaterial: _MeshStandardMaterial
})

createRoot(canvasNode).render(<mesh>
    <boxGeometry />
    <meshStandardMaterial />
  </mesh>);"
`;

exports[`works with animated from @react-spring/three 1`] = `
"import { Color as _Color } from \\"three\\";
import { MeshPhongMaterial as _MeshPhongMaterial } from \\"three\\";
import { BoxGeometry as _BoxGeometry } from \\"three\\";
import { Mesh as _Mesh } from \\"three\\";
import { extend as _extend } from \\"@react-three/fiber\\";
import { useState } from \\"react\\";
import { createRoot } from \\"@react-three/fiber\\";
import { useSpring, animated } from \\"@react-spring/three\\";

_extend({
  Mesh: _Mesh,
  BoxGeometry: _BoxGeometry,
  MeshPhongMaterial: _MeshPhongMaterial,
  Color: _Color
})

function Box(props) {
  const [active, setActive] = useState(false);
  const {
    scale
  } = useSpring({
    scale: active ? 1.5 : 1
  });
  return <animated.mesh scale={scale} onClick={() => setActive(!active)} {...props}>
      <boxGeometry />
      <meshPhongMaterial color=\\"royalblue\\" />
    </animated.mesh>;
}

createRoot(canvasNode).render(<Box />);"
`;

exports[`works with drei 1`] = `
"import { MeshStandardMaterial as _MeshStandardMaterial } from \\"three\\";
import { BoxGeometry as _BoxGeometry } from \\"three\\";
import { Mesh as _Mesh } from \\"three\\";
import { extend as _extend } from \\"@react-three/fiber\\";
import { createRoot } from '@react-three/fiber';
import { Html } from 'drei';

_extend({
  Mesh: _Mesh,
  BoxGeometry: _BoxGeometry,
  MeshStandardMaterial: _MeshStandardMaterial
})

createRoot(canvasNode).render(<mesh>
    <boxGeometry />
    <meshStandardMaterial />
    <Html>
      <div>Hello World</div>
    </Html>
  </mesh>);"
`;

exports[`works with import shadowing 1`] = `
"import { MeshStandardMaterial as _MeshStandardMaterial } from \\"three\\";
import { BoxGeometry as _BoxGeometry } from \\"three\\";
import { Mesh as _Mesh } from \\"three\\";
import { extend as _extend } from \\"@react-three/fiber\\";
import { createRoot } from '@react-three/fiber';
import { BoxGeometry } from 'three';

_extend({
  Mesh: _Mesh,
  BoxGeometry: _BoxGeometry,
  MeshStandardMaterial: _MeshStandardMaterial
})

createRoot(canvasNode).render(<mesh>
    <boxGeometry />
    <meshStandardMaterial />
  </mesh>);"
`;
